packages/flutter/lib/src/foundation/tree_diagnostics_mixin.dart

```diff
@@ -2,9 +2,1438 @@
+  /// or shown. `nullDescription` provides a description of the property value
+  /// if it is null.  If `defaultValue` is specified, the property is omitted if
+  /// `value == defaultValue`.  If specified, `unit` describes the unit for the
+  /// value (e.g. px).
+
+  factory DiagnosticsNode.intProperty(String name, int value, {
+    bool showNull: true,
+    String nullDescription,
+    bool showName: true,
+    String unit,
+    int defaultValue,
+    bool hidden: false,
+  }) {
+    String header;
+    if (value != null)
+      header = unit != null ? '$value$unit' : value.toString();
+    else
+      header = nullDescription.toString();
+
+    return new _LeafDiagnosticsValue(
+      name: name,
+      header: header,
+      object: value,
+      showNull: showNull,
+      showName: showName,
+      defaultValue: defaultValue,
+      hidden: hidden,
+    );
+  }
+
+  // TODO(jacobr): make this a generic method so that debuggers know the type
+  // of the property even when the instance is null.
+
+  /// Property with an arbitrary typed value.
+  ///
+  /// If the default `value.toString()` does not provide an adequate description
+  /// of the object, specify `header` giving a custom description. If `showNull`
+  /// is true, the property is shown even if its `value` is null. `hidden`
+  /// specifies whether the property should be hidden or shown.
+  /// `nullDescription` provides a description of what a null property value
+  /// means. `showSeparator` indicates whether a separator should be placed
+  /// between the property `name` and `value`.
+  factory DiagnosticsNode.objectProperty(String name, Object value, {
+    String header,
+    bool hidden: false,
+    bool showNull: true,
+    String nullDescription,
+    bool showName: true,
+    bool showSeparator: true,
+  }) {
+    if (nullDescription != null && value == null)
+      header = nullDescription;
+
+    return new _LeafDiagnosticsValue(
+      name: name,
+      header: header ?? value.toString(),
+      object: value,
+      hidden: hidden,
+      showNull: showNull,
+      showName: showName,
+      showSeparator: showSeparator,
+    );
+  }
+
+  /// Property where computing its value could throw an exception.
+  ///
+  /// `header` must always be specified so that `computeValue` does not need to
+  /// be called unless the raw property value is explicitly requested.
+  factory DiagnosticsNode.lazyObjectProperty(String name, Object computeValue(), {
+    @required String header,
+    bool hidden: false,
+    bool showNull: true,
+    bool showName: true,
+  }) {
+    return new _LazyLeafDiagnosticsValue(
+      name: name,
+      header: header,
+      computeValue: computeValue,
+      hidden: hidden,
+      showNull: showNull,
+      showName: showName,
+    );
+  }
+
+  // The following 3 constructors are stubs for adding more detailed diagnostics
+  // later.
+  // TODO(jacobr): switch DiagnosticsNode.objectProperty to be generic and
```

Comment:
we should just do this now, so that we don't have to break the API again later when we remove these